### ISSUE TITLE
proposal: introduce an 'io' package

### ISSUE THREAD

**Issue by sbinet** [gonum/gonum#36]
`gonum` should have a (set of) package(s) dedicated to reading and writing files commonly used in science and data science.
These packages should be pure-Go to be included as "subpackages" of the `gonum/gonum` repo.

I propose to start with the following layout:

- `gonum/io/npyio`: read/write of NPy files
- `gonum/io/matfio`: read/write of Matlab files
- `gonum/io/csvio`: read/write of CSV files (with a strong numerical bias)

`npyio` initial code (to be refactored a bit to not have any dependency on `gonum/matrix`) can be the one from `sbinet/npyio`.

`matfio` could also be taken from `sbinet/matfio`.

`csvio` could be taken from `btracey/numcsv`.

The end game is to nurture a set of interfaces that can be widely implemented in the (data) science communities. (e.g. I know of already 3-4 `.npy` files implementations)

**Comment by sbinet** [comment]
@btracey @Kunde21 @chewxy

**Comment by sbinet** [comment]
already existing read/write `.npy` packages:
- https://github.com/kshedden/gonpy
- https://github.com/cquotient/go-npy
- https://github.com/sbinet/npyio
- https://github.com/chewxy/gorgonia [write-npy](https://github.com/chewxy/gorgonia/blob/ee086174c28f3e5bf92d7ab1e13dc31dd6deca0d/tensor/dense_io.go#L41) [read-npy](https://github.com/chewxy/gorgonia/blob/ee086174c28f3e5bf92d7ab1e13dc31dd6deca0d/tensor/dense_io.go#L243)

**Comment by chewxy** [comment]
To kick off this proposal, here's what is required to serialize anything into a npy file, and a go-equivalent type of the value (not complete, WIP):

* dtype :: `string` - (a not-so-exhaustive list of conversions from Go types to Numpy Dtypes can be found [here](https://github.com/chewxy/gorgonia/blob/master/tensor/types.go#L37)). Full list of supported Dtypes by Numpy can be found [here](https://docs.scipy.org/doc/numpy-1.12.0/reference/arrays.dtypes.html)
* data :: `[]byte`
* memory order :: `bool` (true for FORTRAN/ColMajor)
* shape :: `[]int`

**Other Considerations**

* dtypes other than `float64` : I think the `io` package should support this
* npy versions: version 2 allows up to 4GiB of headers... why would anyone need such a large header is beyond me (use cases?)
* npyz: yes or no? (it's easy to write, but difficult to read without unzipping. I've no clue how to read from a zipped stream)

**Comment by chewxy** [comment]
What I like: npyio feels a lot more disciplined and structured than gorgonia's npy writer and reader. gonpy's implementation is surprisingly quite similar to gorgonia's.  I only found a reader for go-npy...

**Comment by sbinet** [comment]
I would envision something like this:

```go
package npyio

// Marshaler is the interface implemented by an object that can marshal itself into a NPy binary form.
//
// MarshalNPy encodes the receiver into a NPy binary form.
type Marshaler interface {
   MarshalNPy() (data []byte, err error)
}

// Unmarshaler is the interface implemented by an object that can unmarshal an NPy binary representation of itself.
//
// UnmarshalNPy must be able to decode the form generated by MarshalNPy.
// UnmarshalNPy must copy the data if it wishes to retain the data after returning.
type Unmarshaler interface {
  UnmarshalNPy(data []byte) error
}
```

where `data []byte` contains the binary representation of the header + the payload. (but sans the `.npy` magic number)

**Comment by btracey** [comment]
A minor comment at this point, but it might be nice to have a `DenseMarshaler` type which wraps `*Dense` and implements all of these io routines. That way we can keep the `Dense` methods mostly about linear algebra and not about saving to 20 different formats.

**Comment by kortschak** [comment]
>A minor comment at this point, but it might be nice to have a `DenseMarshaler` type which wraps `*Dense` and implements all of these io routines. That way we can keep the `Dense` methods mostly about linear algebra and not about saving to 20 different formats.

Yes, please. A futher possibility is that these routines/types be built around `RawMatrixer` rather than explicitly around `*Dense`.

**Comment by chewxy** [comment]
If this were to be a lib, wouldn't it be better that it handles everything? Then all a type would have to do is pass it its relevant params

```go
type Marshaler interface {
    MarshalNPY(dtype string, shape []int, colmajor bool, data []byte) error
}
```

**Comment by kortschak** [comment]
Where is the `data` coming from in that call? Do you mean `Unmarshal`?

I'm with @btracey on this. I don't want to see an arbitrarily long list of data encoding methods on the data types of matrix. The common specification attributes needed for marshaling and unmarshaling can be added (preferably in a single method that returns a struct of attributes or some other way of handling this) that each of the packages in encoding can then use to self configure.

I would like the "gonum.org/v1/gonum/matrix/encoding/npy" encoding signatures to look something like this

```
package encoding

type T1 ... // either config struct or private type implementing an interface for encoding.

type T2 ...

type Structurer interface {
    Structure() T1
}

type StructureSetter interface {
    SetStructure() T2
}
```

package npy

func MarshalMatrix(v encoding.Structurer) ([]byte, error)

func Unmarshal(data []byte, v encoding.StructureSetter) error

type Decoder struct {
    r io.Reader
    ...
}

func (d *Decoder) Decode(v encoding.StructureSetter) error

type Endcoder struct {
    w io.Writer
    ...
}

func (e *Encoder) Encode(v encoding.Structurer) error
```

and additonal helper methods.

**Comment by chewxy** [comment]
Oops. Yes, sorry, unmarshal, not marshal. Ugh not yet had my morning coffee.

**Comment by kortschak** [comment]
>p/s library writers - casting `[]T` to `[]byte` is perfectly acceptable IMO: `bytes := *(*[]byte)(unsafe.Pointer(&a))` where `a` is of `[]T`

This must have a non-unsafe equivalent. We explicitly support appengine, which means we have to have a reflect or other equivalent.

**Comment by chewxy** [comment]
TIL you can't use unsafe in appengine

**Comment by sbinet** [comment]
@kortschak @btracey I guess what is more or less being described in https://github.com/gonum/gonum/issues/36#issuecomment-305653549 is a general protocol for:

- describing a memory layout (a schema of sorts?)
- describing a protocol to get some "flatten" data from that memory layout
- describing a protocol to get some memory layout from some "flatten" data

and a package that ties all of these protocols,
plus, finally, a myriad of packages that can translate from some on-disk data to the "flatten" data (or, directly, to the memory layout.)

IIRC, that's more or less what [parquet](https://parquet.apache.org/documentation/latest/) and [arrow](https://arrow.apache.org/) are targeted at.

but we could indeed start with only an `numpy.dtype`-like schema/descriptor (which is less expressive than the whole `arrow` stuff.)

**Comment by kczimm** [comment]
Has there been any development or further design of this io package? I'm interested in seeing and helping get this package developed.

**Comment by sbinet** [comment]
Sadly, not much work from my end (busy making Go-HEP understand ROOT files, among other things)

Developing 'io' might actually be an interesting Go2 experience report: I think it ties back to how Go should represent n-dim slices.

**Comment by bruth** [comment]
Hi! This issue has not been updated in some time, but I have a use case to perform operations on a ~200G (possibly larger) `.npz` file. One question I have is whether any of the readers support memory mapping to obviate the need for excessive memory use.

Likewise if you have a suggestion on an alternative multi-dimensional on-disk format that is a bit more language agnostic and efficient to work with, I would be happy to hear suggestions, thanks!

**Comment by kortschak** [comment]
WRT memory mapping, this complicates the implementation because we need to then consider the Windows (and other OS) equivalents to mmap - plan9 e.g. doesn't have mmap AFAIK. This means that we need to implement it on `io.ReaderAt` for the default case. Given this, that would be the starting point, with mmap-based implementations coming as optimisations for platforms that support it or operations like it.

**Comment by sbinet** [comment]
The reading part has already been implemented (for Plan9 as well) in x/exp/mmap.

(And I have started something for the writing part on go-mmap/mmap)

Notre that '.npz' suggests the file is compressed (with zip) so the mmap thing couldn't be applied to the whole file, only to the already uncompressed chunks.

Arrow and/or parquet may be good alternatives.
Zarr (zarray) as well (haven't tested it, but I looked at the specs)

HTH.

**Comment by kortschak** [comment]
The plan9 mmap in x/exp is not actually mmapped. It uses `io.ReaderAt`.

**Comment by sbinet** [comment]
yes, it's using the same strategy than the one you sketched out above :)
(and one can "easily" convert an `io.ReaderAt` into an `io.Reader` via an `io.LimitedReader`)

@bruth as for other alternative multi-dimensional on-disk formats, I guess it depends on the `n` in multi-dimensional.
other that come to mind: FITS (used since the 70's in astro), HDF5 (replacing FITS and nDF in some fields)

**Comment by bruth** [comment]
I appreciate the responses. I have a dataset of floats representing similarity scores for a set of N = ~1.9 million entities. So the number of floats is that N^2 / 2. The access pattern are point lookups between two entities, e.g. "what is the sim score between entity A and D".

So I am looking for the ideal on-disk format leveraging compression, but can support this access pattern with loading all the data in memory. A custom format where the data is ordered and chunked into files with an in-memory index file of some kind seems like a straightforward approach, but I prefer standard formats if one exists supporting this use case.
